mirrord/agent/src/main.rs

```diff
@@ -659,37 +628,33 @@ async fn start_agent(args: Args, watch: drain::Watch) -> Result<()> {
     // initialize.
     println!("agent ready - version {}", env!("CARGO_PKG_VERSION"));
 
-    let mut clients = FuturesUnordered::new();
+    let mut clients: JoinSet<ClientId> = JoinSet::new();
 
-    // For the first client, we use communication_timeout, then we exit when no more
-    // no connections.
-    match timeout(
+    // We wait for the first client until `communication_timeout` elapses.
+    let first_connection = timeout(
         Duration::from_secs(args.communication_timeout.into()),
         listener.accept(),
     )
-    .await
-    {
+    .await;
+    match first_connection {
         Ok(Ok((stream, addr))) => {
-            trace!("start -> Connection accepted from {:?}", addr);
-            if let Some(client) = state
-                .new_connection(
-                    stream,
-                    bg_tasks.clone(),
-                    cancellation_token.clone(),
-                    args.base_protocol_version.clone(),
-                )
-                .await?
-            {
-                clients.push(client)
-            };
+            trace!(peer = %addr, "start_agent -> First connection accepted");
+            clients.spawn(state.clone().serve_client_connection(
```

Comment:
Ah, I see now why you want to `clone` state, instead of making a new one on each connection.